options/options.go

```diff
@@ -784,14 +784,73 @@ func (interp *interpreter) interpretOptions(fqn string, element, opts proto.Mess
 	return nil, nil
 }
 
+// isKnownField returns true if the given option is for a known field of the
+// given options message descriptor and will be serialized using the expected
+// wire type for that known field.
 func isKnownField(desc protoreflect.MessageDescriptor, opt *interpretedOption) bool {
 	var num int32
 	if len(opt.pathPrefix) > 0 {
 		num = opt.pathPrefix[0]
 	} else {
 		num = opt.number
 	}
-	return desc.Fields().ByNumber(protoreflect.FieldNumber(num)) != nil
+	fd := desc.Fields().ByNumber(protoreflect.FieldNumber(num))
+	if fd == nil {
+		return false
+	}
+
+	// Before the full wire type check, we do a quick check that will usually pass
+	// and allow us to short-circuit the logic below.
+	if fd.IsList() == opt.repeated && fd.Kind() == opt.kind {
+		return true
+	}
+
+	// We figure out the wire type this interpreted field will use when serialized.
+	var wireType protowire.Type
+	switch {
+	case len(opt.pathPrefix) > 0:
+		// If path prefix exists, this field is nested inside a message.
+		// And messages use bytes wire type.
+		wireType = protowire.BytesType
+	case opt.repeated && opt.packed && canPack(opt.kind):
+		// Packed repeated numeric scalars use bytes wire type.
+		wireType = protowire.BytesType
+	default:
+		switch opt.kind {
+		case protoreflect.StringKind, protoreflect.BytesKind, protoreflect.MessageKind:
+			wireType = protowire.BytesType
+		case protoreflect.GroupKind:
+			wireType = protowire.StartGroupType
+		case protoreflect.Fixed32Kind, protoreflect.Sfixed32Kind, protoreflect.FloatKind:
+			wireType = protowire.Fixed32Type
+		case protoreflect.Fixed64Kind, protoreflect.Sfixed64Kind, protoreflect.DoubleKind:
+			wireType = protowire.Fixed64Type
+		default:
+			// everything else uses varint
+			wireType = protowire.VarintType
+		}
+	}
+
+	// And then we see if the wire type we just determined is compatible with
+	// the field descriptor we found.
+	if fd.IsList() && canPack(fd.Kind()) && wireType == protowire.BytesType {
+		// Even if fd.IsPacked() is false, bytes type is still accepted for
+		// repeated scalar numerics, so that changing a field from
+		return true
+	}
+	switch fd.Kind() {
+	case protoreflect.StringKind, protoreflect.BytesKind, protoreflect.MessageKind:
+		return wireType == protowire.BytesType
+	case protoreflect.GroupKind:
+		return wireType == protowire.StartGroupType
+	case protoreflect.Fixed32Kind, protoreflect.Sfixed32Kind, protoreflect.FloatKind:
+		return wireType == protowire.Fixed32Type
+	case protoreflect.Fixed64Kind, protoreflect.Sfixed64Kind, protoreflect.DoubleKind:
+		return wireType == protowire.Fixed64Type
+	default:
+		// everything else uses varint
+		return wireType == protowire.VarintType
```

Comment:
nit: This looks similar to the code above, and I think we could also use a mapping function from `protoreflect.*Kind` -> `protowire.*Type` in appendOptionBytesSingle to keep this logic all in one place. What do you think about adding a `mapReflectKindToWireType` method?

Reply:
Yep, good call. Done